### PR TITLE
Heading anchors and icons for DO/DO NOT rules.  

### PR DESCRIPTION
**Pull request by adamskt**
- Added jemoji for icons in do/do not sections
- added Pure Liquid heading anchors
- added a link to our Font Awesome kit to get more icons
 #110

### CODE REVIEW

**Comment by adamskt** [comment]
Added bonus: any Github emoji should work in our doc markdown now.

File: docs/coding/csharp.md [hunk 1]
```diff
@@ -3,139 +3,139 @@
...
 Exception: Unless you need to pass a member value by ref, and it cannot be a local variable. Example: Maintaining an internal reference count using Interlocked.
 This largely removes the need for readonly fields in favor of properties.
-- AVOID: Hungarian notation
-- AVOID: incorporating the data type into the name.
-- DO: Use C# alias types rather than their System namespace counterparts (ie. ‘int’ instead of ‘System.Int32’)
-- DO: Use plural names for collections
-- DO: Make boolean names should contain Is/Are/Can/Has which implies Yes/No or True/False values. 
-- AVOID: Double negatives when referring to booleans.
-- AVOID: var in cases where the the type of var is not obvious. Common exceptions would be constructors and generic method calls where it is clear the generic type is the return value.
+- :no_entry: AVOID Hungarian notation
```
Comment (Keboo) [inline docs/coding/csharp.md:32]:
```suggestion
- :no_entry: AVOID: Hungarian notation
```

File: docs/coding/csharp.md [hunk 2]
```diff
@@ -3,139 +3,139 @@
...
+- :no_entry: AVOID incorporating the data type into the name.
```
Comment (Keboo) [inline docs/coding/csharp.md:33]:
```suggestion
- :no_entry: AVOID: incorporating the data type into the name.
```

File: docs/coding/csharp.md [hunk 3]
```diff
@@ -3,139 +3,139 @@
...
+- :heavy_check_mark: DO: Use C# alias types rather than their System namespace counterparts (ie. ‘int’ instead of ‘System.Int32’)
+- :heavy_check_mark: DO: Use plural names for collections
+- :heavy_check_mark: DO: Make boolean names should contain Is/Are/Can/Has which implies Yes/No or True/False values. 
+- :no_entry: AVOID Double negatives when referring to booleans.
```
Comment (Keboo) [inline docs/coding/csharp.md:37]:
```suggestion
- :no_entry: AVOID: Double negatives when referring to booleans.
```

File: docs/coding/csharp.md [hunk 4]
```diff
@@ -3,139 +3,139 @@
...
+- :no_entry: AVOID var in cases where the the type of var is not obvious. Common exceptions would be constructors and generic method calls where it is clear the generic type is the return value.
```
Comment (Keboo) [inline docs/coding/csharp.md:38]:
```suggestion
- :no_entry: AVOID: var in cases where the the type of var is not obvious. Common exceptions would be constructors and generic method calls where it is clear the generic type is the return value.
```

File: docs/coding/csharp.md [hunk 5]
```diff
@@ -3,139 +3,139 @@
...
 
 ### Abbreviations
-- AVOID: Abbreviations unless project approved or generally accepted and documented.
-- DO: If an object property name fully or partially uses a two letter abbreviation, it will be all uppercase; if it is three letters it will be mixed.
+- :no_entry: AVOID Abbreviations unless project approved or generally accepted and documented.
```
Comment (Keboo) [inline docs/coding/csharp.md:41]:
```suggestion
- :no_entry: AVOID: Abbreviations unless project approved or generally accepted and documented.
```

File: docs/coding/csharp.md [hunk 6]
```diff
@@ -3,139 +3,139 @@
...
-- DO: Give source files the name of the class in the file
-- DO: Group members into sections based on type (Constants, Constructors, Fields and Properties, Methods, Interface implementations). Static members should be grouped with their respective instance members. (e.g. static methods grouped with instance methods)
-- AVOID: Lines longer than 150 characters
-- CONSIDER: Group members in a file by their member type. For examples, keep all properties together, all methods together, constructors/finalizers together.
+- :heavy_check_mark: DO: Put a single class per file unless varying arity, code is generated, creating nested classes, or there is a needed  “buddy class” to add DataAnnotations.
+- :heavy_check_mark: DO: Give source files the name of the class in the file
+- :heavy_check_mark: DO: Group members into sections based on type (Constants, Constructors, Fields and Properties, Methods, Interface implementations). Static members should be grouped with their respective instance members. (e.g. static methods grouped with instance methods)
+- :no_entry: AVOID Lines longer than 150 characters
```
Comment (Keboo) [inline docs/coding/csharp.md:63]:
```suggestion
- :no_entry: AVOID: Lines longer than 150 characters
```

File: docs/coding/csharp.md [hunk 7]
```diff
@@ -3,139 +3,139 @@
...
+- :x: DO NOT: Access fields outside of their property wrapper or the constructor.
 Exception: MVVM situations where the property setter is being used for input validation.
-- DO: Use read-only properties (C# 6.0) in favor of read-only fields
+- :heavy_check_mark: DO: Use read-only properties (C# 6.0) in favor of read-only fields
 
 ### Regions
-- AVOID: Using regions that do not add value.
-- CONSIDER: Using regions around interface implementations
-- DO: Provide a name for all regions (specify text after #region)
+- :no_entry: AVOID Using regions that do not add value.
```
Comment (Keboo) [inline docs/coding/csharp.md:73]:
```suggestion
- :no_entry: AVOID: Using regions that do not add value.
```

File: docs/coding/csharp.md [hunk 8]
```diff
@@ -3,139 +3,139 @@
...
-- DO: Specify CultureInfo when calling Parse/TryParse when parsing strings that may be localized. 
+- :heavy_check_mark: DO: Store all dates in UTC. Convert to locale specific values at the client.
+- :heavy_check_mark: DO: Specify CultureInfo when calling Parse/TryParse when parsing strings that may be localized. 
 
 ### Methods
-- DO NOT: Return null from a method with a return type of IEnumerable<T>.
-- AVOID: Null checks on IEnumerable<T> return values.
-- DO NOT: Return null from a method with a return type Task<T>.
+- :x: DO NOT: Return null from a method with a return type of IEnumerable<T>.
+- :no_entry: AVOID Null checks on IEnumerable<T> return values.
```
Comment (Keboo) [inline docs/coding/csharp.md:102]:
```suggestion
- :no_entry: AVOID: Null checks on IEnumerable<T> return values.
```

File: docs/coding/csharp.md [hunk 9]
```diff
@@ -3,139 +3,139 @@
...
-- AVOID: Catching and wrapping in nonspecific exceptions.
-- DO NOT: Throw System.Exception, System.SystemException, or System.NullReferenceException. Throw the existing framework exception instead, such as System.InvalidOperationException. 
+- :heavy_check_mark: DO: Throw the most specific (most derived) exception that makes sense.
+- :x: DO NOT: Swallow errors by catching nonspecific exceptions such as System.Exception and System.SystemException unless the exception is re-thrown
+- :heavy_check_mark: DO: Catch only exceptions you are explicitly handling. Use exception filters to further limit which exceptions you handle.
+- :x: DO NOT: Over catch exceptions. Exceptions should often be allowed to propagate up the call stack. Use exception filters to avoid catching exceptions that you cannot handle.
+- :heavy_check_mark: DO: In a catch statement that throws an exception, throw the original (using throw;)  or wrapped exception if the exception type is not appropriate to re-throw  (throw new Exception(exception)). This maintains the stack location of the original error.
+- :no_entry: AVOID Defining custom exception classes.
```
Comment (Keboo) [inline docs/coding/csharp.md:111]:
```suggestion
- :no_entry: AVOID: Defining custom exception classes.
```

File: docs/coding/csharp.md [hunk 10]
```diff
@@ -3,139 +3,139 @@
...
+- :no_entry: AVOID Returning error codes.  Exceptions are the primary means of reporting errors in frameworks.
```
Comment (Keboo) [inline docs/coding/csharp.md:112]:
```suggestion
- :no_entry: AVOID: Returning error codes.  Exceptions are the primary means of reporting errors in frameworks.
```

File: docs/coding/csharp.md [hunk 11]
```diff
@@ -3,139 +3,139 @@
...
+- :heavy_check_mark: CONSIDER: terminating the process by calling System.Environment.FailFast if code encounters a situation where it is unsafe for further execution.
+- :no_entry: AVOID Creating APIs that when called can result in a system failure.  If such a failure can occur, call Environment.FailFast when the system failure occurs instead.
```
Comment (Keboo) [inline docs/coding/csharp.md:114]:
```suggestion
- :no_entry: AVOID: Creating APIs that when called can result in a system failure.  If such a failure can occur, call Environment.FailFast when the system failure occurs instead.
```

File: docs/coding/csharp.md [hunk 12]
```diff
@@ -3,139 +3,139 @@
...
+- :x: DO NOT: Use exceptions for normal flow of control.
+- :x: DO NOT: Have public members that can either throw or not based on some option.
+- :x: DO NOT: Have public members that return exceptions as return value or an out parameter.
+- :no_entry: AVOID Explicitly throwing exceptions from finally blocks. (Implicitly thrown exceptions resulting from calling methods that throw are acceptable.)
```
Comment (Keboo) [inline docs/coding/csharp.md:118]:
```suggestion
- :no_entry: AVOID: Explicitly throwing exceptions from finally blocks. (Implicitly thrown exceptions resulting from calling methods that throw are acceptable.)
```

File: docs/coding/csharp.md [hunk 13]
```diff
@@ -3,139 +3,139 @@
...
+- :x: DO NOT: Create or throw custom exception types that callers are not expected to handle. Throw one of the existing Framework exceptions instead. 
+- :heavy_check_mark: CONSIDER: Creating and throwing custom exceptions if you have a unique program error that cannot be communicated using an existing framework exceptions
+- :x: DO NOT: Create a new exception type if the exception would not be handled differently than an existing Framework exception – throw the existing framework exception instead.
+- :heavy_check_mark: DO: Use try-finally for clean up work and avoid using try-catch.
+- :heavy_check_mark: CONSIDER: Wrapping specific exceptions thrown from the lower layer in a more appropriate exception if the lower-layer exception does not make sense in the context of the higher-layer operation.
+- :no_entry: AVOID Catching and wrapping in nonspecific exceptions.
```
Comment (Keboo) [inline docs/coding/csharp.md:124]:
```suggestion
- :no_entry: AVOID: Catching and wrapping in nonspecific exceptions.
```

File: docs/coding/csharp.md [hunk 14]
```diff
@@ -3,139 +3,139 @@
...
-- DO: Make exceptions runtime serializable.
-- DO: Provide exception constructors for default, string message, and Exception inner.
-- CONSIDER: Providing exception properties for programmatic access to extra information relevant to the exception.
-- CONSIDER: Providing a tester method for members that might throw exceptions in common scenarios to avoid performance problems (Tester-Doer Pattern).  E.g. TryParse().
-- DO: Provide a Try-X method for every throwing X method.
+- :x: DO NOT: Throw or derive from System.ApplicationException.
+- :heavy_check_mark: DO: Throw an InvalidOperationException if the object is not in a valid state.
+- :heavy_check_mark: DO: Throw ArgumentException or one of its subtypes (ArgumentNullException) if bad arguments are passed to a member.  Prefer the most derived exception type, if applicable. Set the parameter name of the bad argument that was passed. 
+- :x: DO NOT: Explicitly throw ComException, ExecutionEngineException, or SEHException.
+- :no_entry: AVOID Deep exception hierarchies.
```
Comment (Keboo) [inline docs/coding/csharp.md:131]:
```suggestion
- :no_entry: AVOID: Deep exception hierarchies.
```

**Pull request merged**